tpm2/pcr_profile.go

```diff
@@ -467,180 +545,248 @@ func (p *PCRProtectionProfile) AddProfileOR(profiles ...*PCRProtectionProfile) *
-// through the instructions in this profile.
-func (p *PCRProtectionProfile) traverseInstructions() *pcrProtectionProfileIterator {
-	i := &pcrProtectionProfileIterator{}
-	i.descendInToBranches(p.root)
-	return i
+func (c *pcrProtectionProfileExecContext) popNextInstr() pcrProtectionProfileInstr {
+	branch := c.branchStack[0]
+	instr := branch.instrs[0]
+	branch.instrs = branch.instrs[1:]
+	return instr
 }
 
-type pcrProtectionProfileStringifyBranchContext struct {
-	index int
-	total int
+func (c *pcrProtectionProfileExecContext) done() bool {
+	return len(c.branchStack) == 0
 }
 
-func (p *PCRProtectionProfile) String() string {
-	var b bytes.Buffer
-
-	contexts := []*pcrProtectionProfileStringifyBranchContext{{index: 0, total: 1}}
-	branchStart := false
-
-	iter := p.traverseInstructions()
-	for len(contexts) > 0 {
-		fmt.Fprintf(&b, "\n")
-		depth := len(contexts) - 1
-		if branchStart {
-			branchStart = false
-			fmt.Fprintf(&b, "%*sBranch %d {\n", depth*3, "", contexts[0].index)
-		}
+// run executes this profile with the supplied handler.
+func (p *PCRProtectionProfile) run(handler pcrProtectionProfileInstrHandler) error {
+	context := newPcrProtectionProfileExecContext(p, handler)
 
-		switch i := iter.next().(type) {
-		case *pcrProtectionProfileAddPCRValueInstr:
-			fmt.Fprintf(&b, "%*s AddPCRValue(%v, %d, %x)", depth*3, "", i.alg, i.pcr, i.value)
-		case *pcrProtectionProfileAddPCRValueFromTPMInstr:
-			fmt.Fprintf(&b, "%*s AddPCRValueFromTPM(%v, %d)", depth*3, "", i.alg, i.pcr)
-		case *pcrProtectionProfileExtendPCRInstr:
-			fmt.Fprintf(&b, "%*s ExtendPCR(%v, %d, %x)", depth*3, "", i.alg, i.pcr, i.value)
-		case *pcrProtectionProfileBranchPointInstr:
-			contexts = append([]*pcrProtectionProfileStringifyBranchContext{{index: 0, total: len(i.bp.childBranches)}}, contexts...)
-			fmt.Fprintf(&b, "%*s BranchPoint(", depth*3, "")
-			branchStart = true
-		case *pcrProtectionProfileEndBranchInstr:
-			contexts[0].index++
-			if len(contexts) > 1 {
-				// This is the end of a sub-branch rather than the root profile.
-				fmt.Fprintf(&b, "%*s}", depth*3, "")
-			}
-			switch {
-			case contexts[0].index < contexts[0].total:
-				// There are sibling branches to print.
-				branchStart = true
-			case len(contexts) > 1:
-				// This is the end of a branch point. Printing will continue with the parent branch.
-				fmt.Fprintf(&b, "\n%*s )", (depth-1)*3, "")
-				fallthrough
-			default:
-				// Return to the parent branch's context.
-				contexts = contexts[1:]
-			}
+	for !context.done() {
+		if err := context.popNextInstr().run(context); err != nil {
+			return err
 		}
 	}
 
-	return b.String()
+	return nil
 }
 
-// pcrProtectionProfileComputeContext records state used when computing PCR
-// values for a PCRProtectionProfile
-type pcrProtectionProfileComputeContext struct {
-	parent *pcrProtectionProfileComputeContext
-	values pcrValuesList
+type pcrProtectionProfileStringContext struct {
```

Comment:
it is strange to call this a contex, as now it's really a handler, maybe just pcrProtectionProfileStringifier ?